### PR TITLE
change maid_manager_account_decrease_with_churn to use duplicated put

### PR DESCRIPTION
**Pull request by maqi**
This change is [Reviewable](https://reviewable.io/reviews/maidsafe/safe_vault/553)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @Fraser999

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: tests/maid_manager.rs [hunk 1]
```diff
@@ -233,10 +228,17 @@ fn maid_manager_account_decrease_with_churn() {
...
-            client.put(data.clone());
+        if i % 2 == 0 {
+            for data in (0..4).map(|_| {
+                Data::Structured(test_utils::random_structured_data(100000, &full_id, &mut rng))
+            }) {
+                client.put(data.clone());
+                data_list.push(data.clone());
+            }
+        } else {
+            for data in &data_list {
```
Comment (Viv-Rajkumar) [inline tests/maid_manager.rs]: Any specific reason we're trying to repeat failure puts cumulatively? we don't clear `data_list`, so in every odd iteration, we're going to try to PUT all items in `data_list` again, so iteration 9 would be trying 20 PUTs?

We could maybe just try with the last X num of items from `data_list` as was added in the previous iteration?

**Comment by Viv-Rajkumar** [comment]
> So if all puts attempted in this test had succeeded(as in the code this test is testing was functioning wrong)

should be:

So if all MM ignored the put failures from DM to actually decrease store count and returned success/failure to client(as in the code this test is testing was functioning wrong)

**Comment by maqi** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[tests/maid_manager.rs, line 239 at r1](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSlt135Zlz9lXhBhNGD-r1-239:-KSowqe8uoSDshU--vEg:bcmvn9o) ([raw file](https://github.com/maidsafe/safe_vault/blob/b52884606f49aec9248f6d9494f772e3f9a4a6f7/tests/maid_manager.rs#L239)):*
<details><summary><i>Previously, Viv-Rajkumar (Viv Rajkumar) wrote…</i></summary>
> Any specific reason we're trying to repeat failure puts cumulatively? we don't clear `data_list`, so in every odd iteration, we're going to try to PUT all items in `data_list` again, so iteration 9 would be trying 20 PUTs?
> 
> 
> 
> We could maybe just try with the last X num of items from `data_list` as was added in the previous iteration?
> 
>  
</details>
in each iteration, different nodes will added/removed. so the cumulatively repeat of failure puts is just to ensure those old data has been properly relocated, so the account can be updated properly/consistently among all nodes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by maqi** [comment]
what is this about?

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-:-KSozEhME5E-pU2Biwb2:b-9wyecr)*

**Comment by Fraser999** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/maid_manager.rs, line 239 at r1](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSlt135Zlz9lXhBhNGD-r1-239:-KSpsXuWwi2GET-1Q0_p:b-6zwe88) ([raw file](https://github.com/maidsafe/safe_vault/blob/b52884606f49aec9248f6d9494f772e3f9a4a6f7/tests/maid_manager.rs#L239)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> in each iteration, different nodes will added/removed. so the cumulatively repeat of failure puts is just to ensure those old data has been properly relocated, so the account can be updated properly/consistently among all nodes.
</details>
That's probably something that's out of scope for this test though.  We shouldn't have to Put the old data again to work round a problem which should be getting tested in a different test, and hopefully doesn't exist as a problem anyway?

---

*[tests/maid_manager.rs, line 216 at r2](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSpt9hqarG0Y2Y4tytf:-KSpt9hqarG0Y2Y4tytg:b6ie3vg) ([raw file](https://github.com/maidsafe/safe_vault/blob/345535886e42ba51401c6181867e27ad127428f3/tests/maid_manager.rs#L216)):*
> ```Rust
>     let mut event_count = 0;
>     let mut data_list = Vec::new();
>     let mut expect_count = 1;
> ```

We don't need this variable at all, do we?  It always ends up as `41` when we finally assert on it.  I think it would be good to use named variables for the two loop counts and use those in the assertion (e.g. `assert_eq!(count, Some(iterations * chunks_per_iter + 1));`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by Fraser999** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/maid_manager.rs, line 215 at r2](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSpvA_4pV-B7eIFDwfj:-KSpvA_5EYgM5C1ITRrp:b-fowgtk) ([raw file](https://github.com/maidsafe/safe_vault/blob/345535886e42ba51401c6181867e27ad127428f3/tests/maid_manager.rs#L215)):*
> ```Rust
>     let full_id = client.full_id().clone();
>     let mut event_count = 0;
>     let mut data_list = Vec::new();
> ```

I guess we could reduce the scope of this variable by just moving this line inside the loop, then we can get rid of the line calling `clear()` each time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by maqi** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/maid_manager.rs, line 215 at r2](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSpvA_4pV-B7eIFDwfj:-KSqrU_L0TWbkltabpub:b-wbcyw) ([raw file](https://github.com/maidsafe/safe_vault/blob/345535886e42ba51401c6181867e27ad127428f3/tests/maid_manager.rs#L215)):*
<details><summary><i>Previously, Fraser999 (Fraser Hutchison) wrote…</i></summary>
> I guess we could reduce the scope of this variable by just moving this line inside the loop, then we can get rid of the line calling `clear()` each time.
</details>
the scope cannot be reduced, because data_list is being populated in iteration 1, then being read in iteration 2.
so basically: in one iteration we put new data, then in next iteration we put the same data again (MM account will be increased then decreased, as received put_failure from DM)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by Fraser999** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/maid_manager.rs, line 215 at r2](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSpvA_4pV-B7eIFDwfj:-KSuXCb3zEdhn7LVS6VR:b-d7o9am) ([raw file](https://github.com/maidsafe/safe_vault/blob/345535886e42ba51401c6181867e27ad127428f3/tests/maid_manager.rs#L215)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> the scope cannot be reduced, because data_list is being populated in iteration 1, then being read in iteration 2.
> so basically: in one iteration we put new data, then in next iteration we put the same data again (MM account will be increased then decreased, as received put_failure from DM)
</details>
Yeah - you're right - sorry.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by afck** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[tests/maid_manager.rs, line 238 at r3](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSxqnXB6YhaP4hYCSu8:-KSxqnXB6YhaP4hYCSu9:b-x093r5) ([raw file](https://github.com/maidsafe/safe_vault/blob/b16aafa50e28e41887bd25c09538230035d5875a/tests/maid_manager.rs#L238)):*
> ```Rust
>             }) {
>                 client.put(data.clone());
>                 data_list.push(data.clone());
> ```

That second `clone()` is probably not needed anymore. The `data` can be moved into the `data_list`.

---

*[tests/maid_manager.rs, line 242 at r3](https://reviewable.io:443/reviews/maidsafe/safe_vault/553#-KSxrRM3k5azsu3DQ4Yc:-KSxrRM3k5azsu3DQ4Yd:bvb9ds6) ([raw file](https://github.com/maidsafe/safe_vault/blob/b16aafa50e28e41887bd25c09538230035d5875a/tests/maid_manager.rs#L242)):*
> ```Rust
>         } else {
>             for data in &data_list {
>                 client.put(data.clone());
> ```

Please add a comment that this is expected to fail and that therefore it is not expected to increase the `Put` count in the client manager.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by afck** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by Fraser999** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/553)*

**Comment by maqi** [comment]
The maid_manager::maid_manager_account_decrease_with_churn test fails occasionally, for example with RNG seed: [1907805060, 4035237522, 603661178, 3919140052].

This is due to: 
the new joined close node received account refresh AFTER the put_request, and other existing node doesn't send out any refresh of that account due to missing put_success from DM.
the missing put_success from DM is due to : among 8 DM nodes, 4 has full space, so responded with put_failure, 4 has free space so put in pending_list and wait for NAE group_refresh. However, none of them will get enough to accumulate.
Regarding the mis-match among nodes of full status, it may because: only in DM::handle_put, we check only use half of the chunk_store's max_space, in other places (pending writes and get_response for refreshed in data), we don't carry out such check. This opens couple of gaps:
1, when receiving multiple put_requests, they may all pass the 50% check, but actually taking too much
   If some nodes completes pending writes procedure quicker, it may reject some put_requests, meanwhile other still accept it.
2, Refreshed in data has the same issue, and it totally by-passed the 50% check.

This PR only change the testing code to use `duplicate put` to trigger `put_failure response` from DM to MM, which will decrease MM account. However, this doesn't resolve the issue described above.


Currently pending this PR to highlight the issue.
